crates/common/tedge_utils/src/fs_notify.rs

```diff
@@ -0,0 +1,441 @@
+use std::{
+    collections::{HashMap, HashSet},
+    hash::Hash,
+    path::{Path, PathBuf},
+};
+
+use async_stream::try_stream;
+pub use futures::{pin_mut, Stream, StreamExt};
+use inotify::{EventMask, Inotify, WatchMask};
+use strum_macros::Display;
+use try_traits::default::TryDefault;
+
+#[derive(Debug, Display, PartialEq, Eq, Clone, Hash)]
+pub enum Masks {
+    Modified,
+    Deleted,
+    Created,
+    Undefined,
+}
+
+impl From<Masks> for WatchMask {
+    fn from(masks: Masks) -> Self {
+        match masks {
+            Masks::Modified => WatchMask::MODIFY,
+            Masks::Deleted => WatchMask::DELETE,
+            Masks::Created => WatchMask::CREATE,
+            Masks::Undefined => WatchMask::empty(),
+        }
+    }
+}
+
+impl From<EventMask> for Masks {
+    fn from(em: EventMask) -> Self {
+        match em {
+            EventMask::MODIFY => Masks::Modified,
+            EventMask::DELETE => Masks::Deleted,
+            EventMask::CREATE => Masks::Created,
+            _ => Masks::Undefined,
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum NotifyStreamError {
+    #[error(transparent)]
+    FromIOError(#[from] std::io::Error),
+
+    #[error("Error starting fs notification service.")]
+    NotifyInitError,
+
+    #[error("Error creating event stream")]
+    ErrorCreatingStream,
+
+    #[error("Error normalising watcher for: {path:?}")]
+    ErrorNormalisingWatcher { path: PathBuf },
+
+    // FIXME: how to show which mask is unsupported when inotify::WatchMask
+    // does not impl Display?
+    #[error("Unsupported mask.")]
+    UnsupportedWatchMask,
+
+    #[error("Expected watch directory to be: {expected:?} but was: {actual:?}")]
+    WrongParentDirectory { expected: PathBuf, actual: PathBuf },
```

Comment:
I didn't look at the code but this error seems weird and more related to a coding issue. What a user can do with such a message?